lib/pipeline.ts

```diff
@@ -255,25 +259,94 @@ export class Pipeline extends cdk.Construct {
     }));
   }
 
-  private addFailureAlarm(title?: string) {
-    const pipelineFailureTopic = new sns.Topic(this, 'PipelineFailureTopic');
+  private addFailureAlarm(title?: string): cloudwatch.Alarm {
+    return new PipelineWatcher(this, 'PipelineWatcher', {
+      pipeline: this.pipeline,
+      title
+    }).alarm;
+  }
+}
+
+interface PipelineWatcherProps {
+  /**
+   * Code Pipeline to monitor for failed stages
+   */
+  pipeline: cpipeline.Pipeline;
+
+  /**
+   * Set the pipelineName of the alarm description.
+   *
+   * Description is set to 'Pipeline <title> has failed stages'
+   *
+   * @default pipeline's name
+   */
+  title?: string;
+}
+
+/**
+ * Construct which watches a Code Pipeline for failed stages and raises an alarm
+ * if there are any failed stages.
+ *
+ * A function runs every minute and calls GetPipelineState for the provided pipeline's
+ * name, counts the number of failed stages and emits a JSON log { failedCount: <number> }.
+ * A metric filter is then configured to track this value as a CloudWatch metric, and
+ * a corresponding alarm is set to fire when the maximim value of a single 5-minute interval
+ * is >= 1.
+ */
+class PipelineWatcher extends cdk.Construct {
+  public readonly alarm: cloudwatch.Alarm;
+
+  constructor(parent: cdk.Construct, name: string, props: PipelineWatcherProps) {
+    super(parent, name);
+
+    const pipelineWatcher = new lambda.Function(this, 'Poller', {
+      handler: 'index.handler',
+      runtime: lambda.Runtime.NodeJS810,
+      code: lambda.Code.inline(fs.readFileSync(path.join(__dirname, 'pipeline-watcher.js')).toString('utf8')),
+      environment: {
+        pipelineName: props.pipeline.pipelineName
+      }
+    });
+
+    // See https://github.com/awslabs/aws-cdk/issues/1340 for exposing grants on the pipeline.
+    pipelineWatcher.addToRolePolicy(new iam.PolicyStatement()
+      .addResource(props.pipeline.pipelineArn)
+      .addAction('codepipeline:GetPipelineState'));
+
+    // Explicitly and pre-emptively create the function's log group because we need it for the metric filter.
```

Comment:
There's a race condition here between Lambda and CloudFormation. I remember @rix0rrr had to create a custom resource to implement something like that and I wonder if that's the case, if we should actually include this in the Lambda construct library.

Reply:
I can address this by having the EventRule trigger depend on the LogGroup Resource. This ensures the log group is created before the first invocation of the function. Lambda lazily creates the log group on first invocation, so no more race condition. The only down-side I can see is that I have to drop down to L1 and rely on the findChild(‘Resource’) convention.

```typescript
const logGroupResource = logGroup.findChild('Resource') as cdk.Resource;
const triggerResource = trigger.findChild('Resource') as cdk.Resource;
triggerResource.addDependency(logGroupResource);
```

Reply:
I guess... 

We should try to find a better way to allow defining dependencies between resources (https://github.com/awslabs/aws-cdk/issues/95)

We have an issue to try and make log groups accessible in the CDK (https://github.com/awslabs/aws-cdk/issues/667).

Reply:
Another alternative is that I have the handler call `putMetricData` directly instead of using a MetricFilter. Let me know what you think so I can wrap this up.

Reply:
I think the metric filter route is the right one, just want to make sure we have a better path in the CDK for future generations to use it. Tactically, your solution for delivlib is perfectly fine.